### PR TITLE
[SES6] Remove recursion, add custom retries

### PR DESCRIPTION
**Pull request by swiftgist**
bnc:1154096
-----------------

**Checklist:**
- [x] Added unittests and or functional tests
- [ ] Adapted documentation
- [x] Referenced issues or internal bugtracker
- [x] Ran integration tests successfully (trigger with "@susebot run teuthology" in a GitHub comment; see the [wiki](https://github.com/SUSE/DeepSea/wiki/Testing) for more information)

### CODE REVIEW

**Comment by swiftgist** [comment]
The first commit is from #1784

**Comment by swiftgist** [comment]
@susebot run teuthology

**Comment by mgfritch** [comment]
This PR appears to fix the same bnc as PR #1784, maybe it should be a simple backport to the SES5 branch?

Also, this PR is tagged as SES5, but the changes are based on the SES6 branch .. ?

**Comment by jschmid1** [comment]
what's the status here @swiftgist

**Comment by swiftgist** [comment]
I've been waiting to chat with @mgfritch to understand the details.

**Comment by mgfritch** [comment]
@susebot run teuthology

**Comment by mgfritch** [comment]
@susebot run teuthology

**Comment by mgfritch** [comment]
@susebot run teuthology

File: srv/modules/runners/osd.py [hunk 1]
```diff
@@ -157,6 +199,10 @@ def _call(self):
             log.error("No host found")
             return False
 
+        if not self.force:
+            print("Checking if OSD can be destroyed")
+            if self._wait_until_empty() is False:
```
Comment (mgfritch) [inline srv/modules/runners/osd.py]: Minor nit  .. but maybe just `return self._wait_until_empty()` ?

```suggestion
            return self._wait_until_empty()
```

File: tests/unit/runners/test_osd.py [hunk 2]
```diff
@@ -36,56 +37,6 @@ def make_sample_data(operation=operation,
...
-        test_fix = test_fix(operation=operation, force=force)
-        test_fix.remove()
-        test_fix._mark_osd.assert_called_once_with(test_fix, 'out')
-        test_fix._service.assert_called_with(test_fix, 'stop')
-        test_fix._service.assert_any_call(test_fix, 'disable')
-        if not force:
-            test_fix._empty_osd.assert_called_once_with(test_fix)
-        test_fix._purge_osd.assert_called_once_with(test_fix)
-        test_fix._delete_grain.assert_called_once_with(test_fix)
-
```
[the remark this answers is not in the document]
Comment (swiftgist) [reply tests/unit/runners/test_osd.py]: These were related to `_is_empty` which is no longer there.

File: srv/modules/runners/osd.py (hunk 1 above)
Comment (swiftgist) [reply srv/modules/runners/osd.py]: It has to `pass` on True so that the remaining steps complete.

Comment (mgfritch) [reply srv/modules/runners/osd.py]: ah yep, I missed that

**Pull request merged**